File: src/main/org/nlogo/api/TrailDrawerInterface.java
package org.nlogo.api;

public interface TrailDrawerInterface
    extends DrawingInterface {
  void drawLine(double x0, double y0, double x1, double y1,
                Object color, double size, String mode);

  void setColors(int[] colors);

  Object getDrawing();

  boolean sendPixels();

  void sendPixels(boolean dirty);

  void stamp(Agent agent, boolean erase);

  void readImage(java.io.InputStream is) throws java.io.IOException;

  void importDrawing(File file) throws java.io.IOException;

  java.awt.image.BufferedImage getAndCreateDrawing(boolean dirty);

  void clearDrawing();

  void exportDrawingToCSV(java.io.PrintWriter writer);

  void rescaleDrawing();
}

File: netlogo-core/src/main/api/TrailDrawerInterface.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.api

import org.nlogo.core.File

trait TrailDrawerInterface extends DrawingInterface {
  def drawLine(x0: Double, y0: Double, x1: Double, y1: Double, color: AnyRef, size: Double, mode: String): Unit
  def setColors(colors: Array[Int], width: Int, height: Int): Unit
  def getDrawing: AnyRef
  def sendPixels: Boolean
  def sendPixels(dirty: Boolean): Unit
  def stamp(agent: Agent, erase: Boolean): Unit
  @throws(classOf[java.io.IOException])
  def readImage(is: java.awt.image.BufferedImage): Unit
  @throws(classOf[java.io.IOException])
  def readImage(is: java.io.InputStream): Unit
  @throws(classOf[java.io.IOException])
  def importDrawing(is: java.io.InputStream, mimeTypeOpt: Option[String] = None): Unit
  @throws(classOf[java.io.IOException])
  def importDrawing(file: File): Unit
  @throws(classOf[java.io.IOException])
  def importDrawingBase64(base64: String): Unit
  def getAndCreateDrawing(dirty: Boolean): java.awt.image.BufferedImage
  def clearDrawing(): Unit
  def exportDrawingToCSV(writer: java.io.PrintWriter): Unit
  def rescaleDrawing(): Unit
  def isDirty: Boolean
}
